File: app/src/main/java/com/rbppl/to_do_list/ui/tasks/TaskListAdapter.java
package com.rbppl.to_do_list.ui.tasks;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import com.rbppl.to_do_list.data.Task;
import com.rbppl.to_do_list.data.TaskEventListener;
import com.rbppl.to_do_list.databinding.TaskListItemBinding;

public class TaskListAdapter extends ListAdapter<Task, TaskListAdapter.TaskViewHolder> {
    private TaskEventListener taskEventListener;
    private OnClickListener taskEditClickListener;
    private OnClickListener taskDeleteClickListener;

    public TaskListAdapter() {
        super(new TaskDiffCallback());
    }

    public void setTaskEventListener(TaskEventListener listener) {
        taskEventListener = listener;
    }

    public void setTaskEditClickListener(OnClickListener listener) {
        taskEditClickListener = listener;
    }

    public void setTaskDeleteClickListener(OnClickListener listener) {
        taskDeleteClickListener = listener;
    }

    class TaskViewHolder extends RecyclerView.ViewHolder {
        TaskListItemBinding binding;

        TaskViewHolder(TaskListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;

            binding.imageButtonEdit.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    Task task = getItem(position);
                    if (taskEditClickListener != null) {
                        taskEditClickListener.onClick(task);
                    } else if (taskEventListener != null) {
                        taskEventListener.onEditTask(task);
                    }
                }
            });

            binding.imageButtonDelete.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    Task task = getItem(position);
                    if (taskDeleteClickListener != null) {
                        taskDeleteClickListener.onClick(task);
                    } else if (taskEventListener != null) {
                        taskEventListener.onDeleteTask(task);
                    }
                }
            });
        }
    }

    @NonNull
    @Override
    public TaskViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        TaskListItemBinding binding = TaskListItemBinding.inflate(inflater, parent, false);
        return new TaskViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull TaskViewHolder holder, int position) {
        Task task = getItem(position);
        holder.binding.textTitle.setText(task.getTitle());
        holder.binding.textDescription.setText(task.getDescription());
        holder.binding.checkboxIsCompleted.setChecked(task.isCompleted());

        holder.binding.checkboxIsCompleted.setOnCheckedChangeListener((buttonView, isChecked) -> {
            task.setCompleted(isChecked);
            if (taskEventListener != null) {
                taskEventListener.onTaskStatusChanged(task);
            }
        });
    }

    interface OnClickListener {
        void onClick(Task task);
    }
}

File: app/src/main/java/com/rbppl/to_do_list/ui/tasks/TaskListAdapter.kt
package com.rbppl.to_do_list.ui.tasks
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.rbppl.to_do_list.data.Task
import com.rbppl.to_do_list.data.TaskEventListener
import com.rbppl.to_do_list.databinding.TaskListItemBinding
class TaskListAdapter : ListAdapter<Task, TaskListAdapter.TaskViewHolder>(TaskDiffCallback()) {
    private var taskEventListener: TaskEventListener? = null
    fun setTaskEventListener(listener: TaskEventListener) {
        taskEventListener = listener
    }
    private var taskEditClickListener: ((Task) -> Unit)? = null
    private var taskDeleteClickListener: ((Task) -> Unit)? = null
    inner class TaskViewHolder(private val binding: TaskListItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        var taskTitle = binding.textTitle
        var taskDescription = binding.textDescription
        var taskCheckBox = binding.checkboxIsCompleted
        init {
            binding.imageButtonEdit.setOnClickListener {
                val position = adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    val task = getItem(position)
                    taskEditClickListener?.invoke(task)
                    taskEventListener?.onEditTask(task)
                }
            }
            binding.imageButtonDelete.setOnClickListener {
                val position = adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    val task = getItem(position)
                    taskDeleteClickListener?.invoke(task)
                    taskEventListener?.onDeleteTask(task)
                }
            }
        }
    }
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TaskViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = TaskListItemBinding.inflate(inflater, parent, false)
        return TaskViewHolder(binding)
    }
    override fun onBindViewHolder(holder: TaskViewHolder, position: Int) {
        val task = getItem(position)
                holder.taskTitle.text = task.title
                holder.taskDescription.text = task.description
        holder.taskCheckBox.isChecked = task.isCompleted
        holder.taskCheckBox.setOnCheckedChangeListener { _, isChecked ->
            task.isCompleted = isChecked
            taskEventListener?.onTaskStatusChanged(task)
    }
}}
